When looking up `Core`, pass `is_being_declared` as `true`

bricknerb opened the pull request:
There's no need to try and poison `Core`.
This doesn't change tests since `Core` isn't being poisoned anyways.
Part of #4622.

danakj commented on toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:
```diff
@@ -369,6 +369,16 @@ fn F() {
 
+// --- fail_dont_implicitly_poison_core.carbon
+
+// CHECK:STDERR: fail_dont_implicitly_poison_core.carbon:[[@LINE+4]]:9: error: `Core.Bool` implicitly referenced here, but package `Core` not found [CoreNotFound]
```
I find this error to be pretty confusing. I would easily try to fix it by trying to understand why Core is not found - is my installation messed up? Where is Core? But of course the answer is I named something Core *later in the file*.

That said, I think the way we want to resolve this is via https://github.com/carbon-language/carbon-lang/pull/4864/files, which is to make Core a keyword. So the `class Core` would not in fact make a class called Core, it would be an invalid use of the Core keyword. Then we should not even have an error here, right?

bricknerb replied:
AFAIU, actually, the reason `Core` is not found is because this is a `no_prelude` test so `Core` is not loaded.
It's not found similar to other no_prelude tests.
This test is different from other tests that trigger this error because it shows that this doesn't cause `Core` to be poisoned, unlike #4900.

If `Core` is a keyword, we would still have this error.
We might have another error saying that class `Core` cannot be defined because `Core` is a keyword.

danakj replied:
Oh okay, thanks. I misunderstood the relationship to #4900.

The failure looks good then, but worrying about poisoning Core specifically is a transient concern, since we won't be able to write `class Core` once we make it a keyword. At that point would we just remove this test? Do you think it's worth adding still as-is?

bricknerb replied:
Changed to class `r#Core`.

danakj commented on toolchain/check/context.cpp:
```diff
@@ -703,7 +703,7 @@ static auto GetCorePackage(Context& context, SemIR::LocId loc_id,
   auto core_scope_result = context.LookupNameInExactScope(
       loc_id, core_name_id, SemIR::NameScopeId::Package,
-      context.name_scopes().Get(SemIR::NameScopeId::Package));
+      context.name_scopes().Get(SemIR::NameScopeId::Package), true);
```
I do think this is the right change to make - not sure how the keyword impl shakes out but we don't want to poison things.
```suggestion
      context.name_scopes().Get(SemIR::NameScopeId::Package), /*is_being_declared=*/true);
```

But I am not sure how to test this in a way that would survive the keyword change, and welcome thoughts on that.

bricknerb replied:
Done (and changed to `r#Core`).

danakj commented on toolchain/check/testdata/function/declaration/no_prelude/name_poisoning.carbon:
```diff
@@ -369,6 +369,16 @@ fn F() {
+// CHECK:STDERR:
+fn F(x: bool);
+
+class Core {}
```
```suggestion
class r#Core {}
```

Would this test the same thing now - and after the keyword change?

bricknerb replied:
I believe so, done.

jonmeow commented:
There's an open proposal to make `Core` a keyword. I think #4900 is more correct.

zygoloid reviewed (commented):
Perhaps we should stop using name lookup to find the `Core` package. We could instead track the `Core` package specially on the `SemIR::File`.

bricknerb commented:
> Perhaps we should stop using name lookup to find the `Core` package. We could instead track the `Core` package specially on the `SemIR::File`.

This makes sense though I assume there might be complexity cost for special casing `Core`.

danakj commented:
> There's an open proposal to make `Core` a keyword. I think #4900 is more correct.

Yeah, I can't tell which of the two PRs is more correct myself, but maybe we should rebase either one on https://github.com/carbon-language/carbon-lang/pull/4909 at which point `r#Core` won't be poisoned, so the test in here should pass without any code changes?

danakj reviewed (approved):
LGTM

bricknerb commented:
I think other changes made this change different, so I'm moving this to draft.

bricknerb committed:
When looking up `Core`, pass `is_being_declared` as `true` because there's no need to try and poison `Core`.
This doesn't change tests since `Core` isn't being poisoned anyways.
Part of #4622.

bricknerb commented:
Sorry for the force push, I had to redo this change given recent changes.

jonmeow commented:
@bricknerb Given #4884, isn't this change a no-op? I ask because also, isn't setting "is_being_declared" when we're doing a non-declaring name lookup a little confusing for future readers?

bricknerb commented:
> @bricknerb Given #4884, isn't this change a no-op?

Yes!

> I ask because also, isn't setting "is_being_declared" when we're doing a non-declaring name lookup a little confusing for future readers?

The name `is_being_declared` might be misleading. It's currently set to `true` only when called from `LookupNameInDecl()` in a scope. In other words, when the name being looked up is known to be in a specific scope.
That's the case here, we know the scope of `Core`, so it makes more sense to pass `true`.
WDYT?

jonmeow commented:
> > @bricknerb Given #4884, isn't this change a no-op?
> 
> Yes!
> 
> > I ask because also, isn't setting "is_being_declared" when we're doing a non-declaring name lookup a little confusing for future readers?
> 
> The name `is_being_declared` might be misleading. It's currently set to `true` only when called from `LookupNameInDecl()` in a scope. In other words, when the name being looked up is known to be in a specific scope. That's the case here, we know the scope of `Core`, so it makes more sense to pass `true`. WDYT?

I think that `Core` is a top-level name, and should pretty much always be there (even before #4884, poisoning would have rarely run and could have poisoned at most one namespace). This is not declaring it, so it will only confuse readers to set "is_being_declared". This PR should be closed, not merged (to be clear, this is what I also meant when I said #4900 is more correct).

jonmeow commented:
BTW, LookupNameInDecl sets `is_being_declared` to true because it's part of declaring names. I think the name is correct.

danakj commented:
> > > @bricknerb Given #4884, isn't this change a no-op?
> > 
> > 
> > Yes!
> > > I ask because also, isn't setting "is_being_declared" when we're doing a non-declaring name lookup a little confusing for future readers?
> > 
> > 
> > The name `is_being_declared` might be misleading. It's currently set to `true` only when called from `LookupNameInDecl()` in a scope. In other words, when the name being looked up is known to be in a specific scope. That's the case here, we know the scope of `Core`, so it makes more sense to pass `true`. WDYT?
> 
> I think that `Core` is a top-level name, and should pretty much always be there (even before #4884, poisoning would have rarely run and could have poisoned at most one namespace). This is not declaring it, so it will only confuse readers to set "is_being_declared". This PR should be closed, not merged (to be clear, this is what I also meant when I said #4900 is more correct).

Hm yeah, when I reviewed earlier it was just a test showing that using Core implicitly did not prevent `r#Core` from being used. Verifying `r#Core` works is good. But I agree we don't need to pass is_being_declared for Core - it's not declaring it, it's a keyword.

bricknerb commented:
> BTW, LookupNameInDecl sets `is_being_declared` to true because it's part of declaring names. I think the name is correct.

Sure, in `LookupNameInDecl()` context it's correct, just not sure that it would stay correct in `LookupNameInExactScope()` context for any (future?) call, given how it's used.
I guess declaring and poisoning together does make sense in principal.

bricknerb commented:
> > > @bricknerb Given #4884, isn't this change a no-op?
> > 
> > 
> > Yes!
> > > I ask because also, isn't setting "is_being_declared" when we're doing a non-declaring name lookup a little confusing for future readers?
> > 
> > 
> > The name `is_being_declared` might be misleading. It's currently set to `true` only when called from `LookupNameInDecl()` in a scope. In other words, when the name being looked up is known to be in a specific scope. That's the case here, we know the scope of `Core`, so it makes more sense to pass `true`. WDYT?
> 
> I think that `Core` is a top-level name, and should pretty much always be there (even before #4884, poisoning would have rarely run and could have poisoned at most one namespace). This is not declaring it, so it will only confuse readers to set "is_being_declared". This PR should be closed, not merged (to be clear, this is what I also meant when I said #4900 is more correct).

SG, closing.
